source/common/http/header_map_impl.h

```diff
@@ -355,10 +499,32 @@ template <class T> std::unique_ptr<T> createHeaderMap(const HeaderMap& rhs) {
   // a few places when dealing with gRPC headers/trailers conversions so it's not trivial to remove.
   // We should revisit this to figure how to make this a bit safer as a non-intentional conversion
   // may have surprising results with different O(1) headers, implementations, etc.
-  auto new_header_map = std::make_unique<T>();
+  auto new_header_map = T::create();
   HeaderMapImpl::copyFrom(*new_header_map, rhs);
   return new_header_map;
 }
 
+struct EmptyHeaders {
+  RequestHeaderMapPtr request_headers = RequestHeaderMapImpl::create();
+  ResponseHeaderMapPtr response_headers = ResponseHeaderMapImpl::create();
+  ResponseTrailerMapPtr response_trailers = ResponseTrailerMapImpl::create();
+};
+
+using StaticEmptyHeaders = ConstSingleton<EmptyHeaders>;
+
+class HeaderMapImplUtility {
+public:
+  struct HeaderMapImplInfo {
+    std::string name_;
```

Comment:
Document what sort of name this is? I'm not clear on the topology of where this is used.